flambe/task/training.py

```diff
@@ -0,0 +1,447 @@
+
+        Returns
+        -------
+        nn.Module
+            The model to train.
+
+        """
+        pass
+
+    @abstractmethod
+    def build_optimizers(self, model: nn.Module) -> Dict[str, Optimizer]:
+        """Build the model, containing all parameters to train.
+
+        Parameters
+        ----------
+        model: nn.Module
+            The model to train
+
+        Returns
+        -------
+        List[Optimizer]
+            A list of optimizers to call on every batch
+
+        """
+        pass
+
+    @abstractmethod
+    def build_schedulers(self,
+                         optimizers: Dict[str, Optimizer],
+                         mode: str = 'iter') -> Dict[str, _LRScheduler]:
+        """Build the model, containing all parameters to train.
+
+        Parameters
+        ----------
+        model: nn.Module
+            The model to train
+
+        Returns
+        -------
+        List[Optimizer]
+            A list of optimizers to call on every batch
+
+        """
+        pass
+
+    @abstractmethod
+    def sample_batches(self, split: str = 'train', train: bool = True) -> Iterable[Batch]:
+        """Get an iterable of batches of data.
+
+        This method is used to create batches of data from an
+        iterable of examples. Both the input examples and the
+        output batches can take any arbitary form.
+
+        Parameters
+        ----------
+        train: bool
+            Whether the batches are meant for training or evaluation.
+            This is important as it defines the input to the
+            ``batch_loss`` and ``batch_metrics`` methods.
+
+        Returns
+        -------
+        Iterable[Batch]
+            An iterable of batches, each of which will be passed to
+            either the ``batch_loss`` or ``batch_metrics`` methods.
+
+        """
+        pass
+
+    @abstractmethod
+    def train_step(self, model: nn.Module, batch: Batch) -> torch.Tensor:
+        """Compute loss on the given batch during training.
+
+        Given a batch, this method computes a training step
+        by providing the output loss. The loss should be returned
+        as a tensor of size B (i.e batch size). This is to ensure
+        correct computation when using gradient accumulation steps.
+
+        ``Important``: this method should *NOT* call the backward
+        method, as this is generally done in the object using the model.
+
+        Parameters
+        ----------
+        model: nn.Module
+            The model.
+        batch: Batch
+            A batch of data to train over. The batch can take any form.
+
+        Returns
+        -------
+        torch.Tensor
+            Should contain at least one element being the loss for the
+            batch. The recommended default key is ``loss``.
+
+        """
+        pass
+
+    @abstractmethod
+    def val_step(self, model: nn.Module, batches: Iterable[Batch]) -> Dict[str, float]:
```

Comment:
If these methods would be objects that are received by the training, then that would help code reuse. For example, if several tasks share a way of doing validation steps, each of them needs to inherit and override, but it will be tricky to share that code among all those clients.

Reply:
Every single one being objects would be overkill and require inspecting many objects to get going, which is against the spirit of this change (you're essentially describing the status quo :p ). but one alternative is to have Training receive a Task object with the methods

Reply:
Yes, but the problem is going to be when you try to test this. In order to test an override, since probably the implementations will depend on the class internal state, you will need to exercise a lot of code from the base class. Probably the thing here is that this class handles to many things

Reply:
yeah that's a fair point. But I think there is a good middle ground we can find :)